### DISCUSSION TITLE
RAG does not seem to work or I use it the wrong way

### DISCUSSION THREAD

**Discussion by qiulang** [open-webui/open-webui#5002] [Q&A]
I upload some documents to the Workspace.  When I  access them using the # symbol before a query, RAG works as I expected and generated the answer from my specified document. But If I do not  specify a document and ask my question, RAG won't work.

The problem is I do not always know which document contains the answer, so I can't always specify a document before asking my question. I had hoped after uploading my documents RAG can just find the answer from them, after all that is how RAG works.

So did I use RAG the wrong way or I missed something here.

An example to show what I meant.

I asked the question "Has that particular company ever  obtained some patents ?" and I happened to know there was a pdf document specifically talked about that, so I specified it and RAG gave me the correct the answer. 

![patents answer](https://github.com/user-attachments/assets/5c5a13f4-3680-4187-a8ac-38f9da977686)

If I didn't not specified a document and asked the same question, RAG said I didn't know.
![I didn't know](https://github.com/user-attachments/assets/86beacad-b11f-441c-8003-d447911f057e)

**Comment by qiulang** [comment]
If I choose "All Documents" I can see from console,  the server scan all the documents again (When I upload those documents in workspace and click scan, it already scanned them). But "All Documents" gave a poor answer compared to the specified document. It said "The company obtained two patents but I do not know the details about them" while the detail is in that document and the specified document gave the correct details!

![All documents](https://github.com/user-attachments/assets/7efdde92-1cef-4f91-a130-b012a31da543)

Now I am inclined to believe this is a bug so I opened an issue against it https://github.com/open-webui/open-webui/issues/5004

**Comment by qiulang** [comment]
I have tried the suggestion in https://github.com/open-webui/open-webui/discussions/3868

1. First ask a question to LLM
2. To # to pick a document and ask a question about it.
3. Ask some question I know the answer lies in the documents I uploaded, RAG still fails to work.

I run out of solution except for downgrading from 0.3.8 to 0.2.0 because I doubt it will work

**Reply by silentoplayz** [reply]
I've noticed that certain models will almost always fail for RAG or doesn't work *at all* with RAG. The `Qwen` family of models aren't RAG-ready and isn't able to tell me about any of my documents, personally. Try swapping the model to Llama 3.1 for example and regenerating the response after you have asked a question about a document using Qwen2 7B. The difference is night and day.

**Reply by qiulang** [reply]
The result llama got me was even worse that is why I switched to Qwen. For example, even when I use # to point the document that has the answer, llama does not give the correct answer. Its answer is much worse than Qwen gave. I attached a screenshot for your reference .

But I am more confused with how RAG work here ? If I need to point to the specific document every time, that is not RAG, is it ?

<img width="913" alt="worse answer" src="https://github.com/user-attachments/assets/4e7e53c3-1209-4dbe-b0da-09276c7f9fca">

**Comment by thiswillbeyourgithub** [comment]
If I may, do a smaller test : find a small text file, send it to the LLM and ask "what type of document is this?

I used that to figure out that for some reason the context in my rag always stays empty. This might be the same issue as you. I have tried switching from ollama to openai and it doesn't change anything. More over looking at the log I see that the document is correctly parsed and embeddings are indeed created.

I referred to this in #5034

**Reply by thiswillbeyourgithub** [reply]
Here's a picture illustrating the issue. We also notice that the rag template is set in the system prompt whereas it should probably be set in the user message. I'm using litellm self hosted btw. I tried disabling all the pipes and filters, switching models etc. Nothing worked

![Screenshot_20240830-152723_1](https://github.com/user-attachments/assets/415901db-13af-41bc-82a7-609dca304720)

**Reply by qiulang** [reply]
@thiswillbeyourgithub

Thanks for the information, but this was I did to debug the issue. I deleted all my old documents, reset vector storage. Then I uploaded  just one article and let RAG scan the article.

The article is this one https://edition.cnn.com/2024/08/28/tech/pavel-durov-telegram-custody-released-intl/index.html  I save it to a pdf and I can see from console, RAG has parsed it,

```
INFO:     127.0.0.1:60126 - "GET /rag/api/v1/config HTTP/1.1" 200 OK
INFO:apps.rag.main:store_data_in_vector_db [Document(metadata={'source': '/Users/langqiu/anaconda3/lib/python3.11/site-packages/data/docs/Telegram boss Pavel Durov faces formal investigation.pdf', 'page': 0, 'start_index': 0}, page_content='Te l e g r a m  fo u n d e r  Pa ve lDurov under formalinvestigation with bail set at$5.56 million, prosecutorsaysParis CNN \xa0—\xa0 noneTelegram founder Pavel Durov is under formalinvestigation and will not be allowed to leave France
...
INFO:apps.rag.main:Updating embedding model: BAAI/bge-base-zh-v1.5 to BAAI/bge-base-zh-v1.5
INFO:config:Saving 'RAG_EMBEDDING_MODEL' to config.json
INFO:sentence_transformers.SentenceTransformer:Load pretrained SentenceTransformer: /Users/langqiu/.cache/huggingface/hub/models--BAAI--bge-base-zh-v1.5/snapshots/f03589ceff5aac7111bd60cfc7d497ca17ecac65
INFO:apps.rag.main:Updating text settings: engine='' tika_server_url='http://tika:9998'
INFO:apps.rag.main:Updating reranking model: BAAI/bge-reranker-v2-m3 to BAAI/bge-reranker-v2-m3
INFO:config:Saving 'RAG_RERANKING_MODEL' to config.json
```

But when I said the question "what happened to Telegram founder in France ?" rag failed to give an correct answer.
The answer  "**As of my knowledge update in September 2023...**" just shows that RAG did not kick in at all.

<img width="1037" alt=" Telegram founder" src="https://github.com/user-attachments/assets/d773943a-c311-4e9b-a393-8e25a0e7f6e7">

I have to point to that pdf to make it work. From console I can see RAG scanned it again.

But that was my question because I don't think that is how rag is supposed to work.

<img width="868" alt="Telegram founder" src="https://github.com/user-attachments/assets/6ecfd7bb-0d79-480b-855b-166f1935866c">

**Reply by qiulang** [reply]
@silentoplayz please check my update and give me some advice how to further debug this issue.

**Reply by silentoplayz** [reply]
> "*When I access them using the # symbol before a query, RAG works as I expected and generated the answer from my specified document. But If I do not specify a document and ask my question, RAG won't work.*"

What you're pointing out to me doesn't seem to be a bug, like at all. If you ask a question within a query and send that to a large language model in a chat **without** providing any document(s) to the model somehow, it won't use the document as knowledge to answer from using RAG. If you want a model to always have context about any documents of yours, you can select the document(s) as `Knowledge` within the model's settings in the `Workspace` -> `Models` section of Open WebUI.
![image](https://github.com/user-attachments/assets/52b3674b-1993-44db-b6fb-89b807492726)

**Reply by silentoplayz** [reply]
Create or edit a modelfile from the `Models` section in the Workspace.
![image](https://github.com/user-attachments/assets/0a5729a7-37ea-4b82-b656-ac3dd55eb465)

**Reply by qiulang** [reply]
Yes I just figured that out, thanks. But I find that if I choose "all documents" as the knowledge, the result I got is not as good as I use # to specify a particular document.

Please refer to the following picture, (just the same as when I first posted my pictures)

<img width="1005" alt="wrong answer" src="https://github.com/user-attachments/assets/db286fa2-420a-4fd9-b15d-0741b5cb138b">

Do I need to add system prompt to make RAG work better ?

**Reply by qiulang** [reply]
llama's answer is purely hallucination even with knowledge set.

<img width="1160" alt="ollama" src="https://github.com/user-attachments/assets/65e97092-1988-4e0d-9ee4-b314f929462e">

I still believe this is a bug.

And why dos Citations show "All documents" twice ?  From the console, it seems that it scans the all documents again.

**Reply by silentoplayz** [reply]
Related - https://github.com/open-webui/open-webui/discussions/3868#discussioncomment-10782274

**Comment by khalilxg** [comment]
same here
